src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -259,12 +258,64 @@ uct_rdmacm_cm_ep_create_reserved_qpn(uct_rdmacm_cm_ep_t *cep,
         ucs_list_add_tail(&ctx->blk_list, &blk->entry);
     }
 
-    cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset;
+    if (ctx->reuse_qpn) {
+        status = uct_rdmacm_cm_get_peer_dev_ctx(ctx, &cep->id->route,
+                                                &peer_dev_ctx);
+        if (status != UCS_OK) {
+            goto out;
+        }
+
+        if (peer_dev_ctx->ref_qpn_blk == NULL) {
+            blk                                 = ucs_list_next(&ctx->blk_list,
+                                                                uct_rdmacm_cm_reserved_qpn_blk_t, entry);
+            peer_dev_ctx->ref_qpn_blk           = blk;
+            peer_dev_ctx->next_avail_qpn_offset = 0;
+        } else if (peer_dev_ctx->next_avail_qpn_offset >= qpns_per_obj) {
+            if (peer_dev_ctx->ref_qpn_blk->entry.next == &ctx->blk_list) {
+                status = uct_rdmacm_cm_reserved_qpn_blk_alloc(
+                        ctx, cep->id->verbs, UCS_LOG_LEVEL_ERROR, &blk);
+                if (status != UCS_OK) {
+                    goto out;
+                }
+
+                ucs_list_add_tail(&ctx->blk_list, &blk->entry);
+            }
+            blk                                 = ucs_list_next(&peer_dev_ctx->ref_qpn_blk->entry,
+                                                                uct_rdmacm_cm_reserved_qpn_blk_t, entry);
+            peer_dev_ctx->ref_qpn_blk           = blk;
+            peer_dev_ctx->next_avail_qpn_offset = 0;
+        } else {
+            blk = peer_dev_ctx->ref_qpn_blk;
+        }
+    } else {
+        blk = ucs_list_tail(&ctx->blk_list, uct_rdmacm_cm_reserved_qpn_blk_t,
+                            entry);
+        if (blk->next_avail_qpn_offset >= qpns_per_obj) {
+            status = uct_rdmacm_cm_reserved_qpn_blk_alloc(ctx, cep->id->verbs,
+                                                          UCS_LOG_LEVEL_ERROR,
+                                                          &blk);
+            if (status != UCS_OK) {
+                goto out;
+            }
+
+            ucs_list_add_tail(&ctx->blk_list, &blk->entry);
+        }
```

Comment:
code duplication

Reply:
@evgeny-leksikov Yes, it's duplicated. Please review the PR as we've talked in the teams. I've added macro to solve this problem